### PR TITLE
Refactor Direct Keymanager to One-Shot Decrypt

### LINKED ISSUE #6804: Huge memory amount used by validator
# 🐞 Bug Report

### Description

Validator uses lots of ram, about 2-3GB for me.
Memory usage is gradually reduced if system memory gets full as used by other processes (eth1 node for example).
After a while memory usage is normal.

### Has this worked before in a previous version?

Don't know. Did not use accounts-v2 until now.

## 🔬 Minimal Reproduction

System memory: 4gb
* Start beacon
* Start a validator with accounts-v2 "direct" wallet with 14 keys in it.
* Observe memory usage
* Start eth1 node
* Observer memory usage

## 🔥 Error

No error messages

## 🌍  Your Environment

**Operating System:**

<pre>
  <code>
Debian 10
  </code>
</pre>

**What version of Prysm are you running? (Which release)**

<pre>
  <code>
1.0.0-alpha.17
  </code>
</pre>

**Anything else relevant (validator index / public key)?**

### LINKED ISSUE #6812: FATAL accounts-v2: Could not list accounts: could not initialize keymanager
# 🐞 Bug Report

### Description

Listing accounts or starting the validators crashes if there was an account saved before

### Has this worked before in a previous version?

N/A

## 🔬 Minimal Reproduction

- Install new Prysm client on a server that already used Prysm
- Import accounts v2 (No error was thrown) ```./prysm.sh validator accounts-v2 import --keys-dir=$HOME/eth2.0-deposit-cli/validator_keys```
- List accounts or run validator

## 🔥 Error

<pre><code>
FATAL accounts-v2: Could not list accounts: could not initialize keymanager: could not initialize direct keymanager: could not initialize keys cache: could not read password for account jointly-perfect-chimp: could not read /root/.eth2validators/prysm-wallet-v2-passwords/jointly-perfect-chimp.pass: open /root/.eth2validators/prysm-wallet-v2-passwords/jointly-perfect-chimp.pass: no such file or directory
</code></pre>

Running multiple times produces a different error:

<pre><code>
panic: send on closed channel

goroutine 94 [running]:
github.com/prysmaticlabs/prysm/shared/mputil.Scatter.func1(0x4000820240, 0x40006ce040, 0x4000203980, 0x4000203920, 0x0, 0x1)
    shared/mputil/scatter.go:45 +0x7c
created by github.com/prysmaticlabs/prysm/shared/mputil.Scatter
    shared/mputil/scatter.go:42 +0x178
</code></pre>

## 🌍  Your Environment

**Operating System:**

<pre>
  <code>
                          ./+o+-       ethereum@ethnode-0b9def7ee
                  yyyyy- -yyyyyy+      OS: Ubuntu 20.04 focal
               ://+//////-yyyyyyo      Kernel: aarch64 Linux 5.4.0-1015-raspi
           .++ .:/++++++/-.+sss/`      Uptime: 4h 33m
         .:++o:  /++++++++/:--:/-      Packages: 716
        o:+o+:++.`..```.-/oo+++++/     Shell: bash 5.0.17
       .:+o:+o/.          `+sssoo+/    Disk: 9.2G / 67G (15%)
  .++/+:+oo+o:`             /sssooo.   CPU: BCM2835 @ 4x 1.5GHz
 /+++//+:`oo+o               /::--:.   RAM: 743MiB / 7811MiB
 \+/+o+++`o++o               ++////.  
  .++.o+++oo+:`             /dddhhh.  
       .+.o+oo:.          `oddhhhh+   
        \+.++o+o``-````.:ohdhhhhh+    
         `:o+++ `ohhhhhhhhyo++os:     
           .o:`.syhhhhhhh/.oo++o`     
               /osyyyyyyo++ooo+++/    
                   ````` +oo+++o\:    
                          `oo++.      

  </code>
</pre>

**What version of Prysm are you running? (Which release)**

<pre>
  <code>
v1.0.0-alpha.17 Running the Prysm startup script
  </code>
</pre>

**Anything else relevant (validator index / public key)?**

The problem is solved by removing the .eth2validators folder and importing again the account

### PR DESCRIPTION
**Pull request by rauljordan**
**What type of PR is this?**

> Feature

**What does this PR do? Why is it needed?**

This PR refactors our direct keymanager implementation to instead store all accounts into a single keystore.json file encrypted by a strong wallet password. This keystore file can simply be decrypted once at runtime, eliminating any memory usage or performance hit we had before. This PR also provides a "silent migration" strategy for users that already created a wallet with Prysm accounts-v2, which will only run once upon startup.

**Which issues(s) does this PR fix?**

Fixes #6804 and fixes #6812 

**Other notes for review**

The accounts are now stored in an `all-accounts.keystore-%d.json` file in the wallet, and passwords are no longer stored on disk, as they can be simply decrypted from the wallet password at runtime.

### CODE REVIEW

File: validator/accounts/v2/accounts_export.go [hunk 1]
```diff
@@ -12,10 +12,10 @@ import (
 	"github.com/logrusorgru/aurora"
 	"github.com/manifoldco/promptui"
 	"github.com/pkg/errors"
+	"github.com/urfave/cli/v2"
+
```
Comment (0xKiwi) [inline validator/accounts/v2/accounts_export.go:16]:
```suggestion
```

File: validator/accounts/v2/accounts_list.go [hunk 2]
```diff
@@ -5,16 +5,16 @@ import (
 	"fmt"
 	"path/filepath"
 
-	"github.com/dustin/go-humanize"
 	"github.com/logrusorgru/aurora"
 	"github.com/pkg/errors"
+	"github.com/urfave/cli/v2"
+
```
Comment (0xKiwi) [inline validator/accounts/v2/accounts_list.go:11]:
```suggestion
```

File: validator/keymanager/v2/direct/direct.go [hunk 3]
```diff
@@ -149,18 +167,12 @@ func (dr *Keymanager) ValidatingAccountNames() ([]string, error) {
...
-		return "", errors.Wrap(err, "could not write password to disk")
-	}
-	// Generates a new EIP-2335 compliant keystore file
-	// from a BLS private key and marshals it as JSON.
-	encoded, err := dr.generateKeystoreFile(validatingKey, password)
+	accountName := petnames.DeterministicName(validatingKey.PublicKey().Marshal(), "-")
+	dr.accountsStore.PrivateKeys = append(dr.accountsStore.PrivateKeys, validatingKey.Marshal())
+	dr.accountsStore.PublicKeys = append(dr.accountsStore.PublicKeys, validatingKey.PublicKey().Marshal())
+	newStore, err := dr.createAccountsKeystore(ctx, dr.accountsStore.PrivateKeys, dr.accountsStore.PublicKeys)
 	if err != nil {
```
Comment (0xKiwi) [inline validator/keymanager/v2/direct/direct.go:174]: Can you add a comment on why you don't return an error here?

File: validator/keymanager/v2/direct/direct.go [hunk 4]
```diff
@@ -270,167 +266,207 @@ func (dr *Keymanager) Sign(ctx context.Context, req *validatorpb.SignRequest) (b
...
-	progressChan := make(chan struct{}, len(accountNames))
-	go func() {
-		defer close(progressChan)
-		var itemsReceived int
-		for range progressChan {
-			itemsReceived++
-			if err := bar.Add(1); err != nil {
-				log.WithError(err).Debug("Could not increase progress bar")
+// ImportKeystores into the direct keymanager from an external source.
+func (dr *Keymanager) ImportKeystores(cliCtx *cli.Context, keystores []*v2keymanager.Keystore) error {
```
Comment (0xKiwi) [inline validator/keymanager/v2/direct/direct.go:270]: This function is a bit large. Can you maybe split it up a bit?

File: validator/keymanager/v2/direct/import.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
+package direct
```
Comment (rauljordan) [inline validator/keymanager/v2/direct/import.go:1]: Review carefully

File: validator/keymanager/v2/direct/migrate.go [hunk 6]
```diff
@@ -0,0 +1,105 @@
+package direct
```
Comment (rauljordan) [inline validator/keymanager/v2/direct/migrate.go:1]: Review carefully

**Comment by prestonvanloon** [comment]
One thing I noticed is that the "pubkey" field in the new accounts.json is blank.
```json
        "uuid": "e4058eba-76b7-4096-bb56-2e302ec16a7a",
        "pubkey": "",
        "version": 4,
        "name": "keystore"
```

I would like to see a pubkeys array, but maybe we should follow up in another PR.

**Pull request merged**